### ISSUE TITLE
调用paddle.grad计算梯度出错 Using paddle.grad to compute gradients raises errors

### ISSUE THREAD

**Issue by holyseven** [PaddlePaddle/Paddle#42685]
### bug描述 Describe the Bug

bug 复现代码，paddel 2.2.2

```python
from paddle.vision.models import resnet50
import numpy as np
import paddle 
from paddle.autograd import grad

paddle_model = resnet50(pretrained=True)
paddle_model.eval()

r = np.random.random((1, 3, 32, 32))
random_image = paddle.to_tensor(r).cast("float32")
l = paddle_model(random_image)

print(grad(l[0], paddle_model.parameters()))  # set only_inputs=True does not help.
```

![image](https://user-images.githubusercontent.com/13829174/167796436-eb7b864c-4d34-44e0-ba4a-8ecd7af4d837.png)

NLP Ernie 模型也一样报错
```python
from paddlenlp.transformers import ErnieForSequenceClassification, ErnieTokenizer
import numpy as np
import paddle
from paddle.autograd import grad

MODEL_NAME = "ernie-1.0"
model = ErnieForSequenceClassification.from_pretrained(MODEL_NAME, num_classes=32)
random_inputs = paddle.to_tensor(np.array([[0]*200]))
l = model(random_inputs, random_inputs)
grad(l, model.parameters())
```

![image](https://user-images.githubusercontent.com/13829174/167796711-0671f299-02de-442b-b455-b77fe62deb17.png)

**Comment by holyseven** [comment]
也许和 #42420 相关

**Comment by JiabinYang** [comment]
应该不是一个问题 你遇到的这个问题的是因为相应的参数没有梯度

**Comment by holyseven** [comment]
通过这个看吗？都是`False`

```python
for n, v in paddle_model.named_parameters():
    print(n, v.stop_gradient)
```

**Comment by JiabinYang** [comment]
> 通过这个看吗？都是`False`
> 
> 
> 
> ```python
> 
> for n, v in paddle_model.named_parameters():
> 
>     print(n, v.stop_gradient)
> 
> ```
- 对于第一个case中，可以看到报错的这个参数是不参与反向提督更新的，在使用`paddle.grad`时需要确保你需要进入计算的input都是支持微分的且需要`stop_gradient=False`
```
from paddle.vision.models import resnet50
import numpy as np
import paddle 
from paddle.autograd import grad

paddle_model = resnet50(pretrained=True)
paddle_model.eval()

r = np.random.random((1, 3, 32, 32))
random_image = paddle.to_tensor(r).cast("float32")
l = paddle_model(random_image)

# print(grad(l[0], paddle_model.parameters()))  # set only_inputs=True does not help.

for v in paddle_model.parameters():
    print("{}, {}".format(v.name, v.stop_gradient))

print(grad(l[0], paddle_model.parameters()))  # set only_inputs=True does not help.
```
![image](https://user-images.githubusercontent.com/22361972/168023257-e0b83c10-b80d-4429-8a8e-3ac69ca62a3d.png)
![image](https://user-images.githubusercontent.com/22361972/168023203-0333d5e4-7fc3-41a6-9518-8fd4e9e15e37.png)

当你把代码改成：
```
from paddle.vision.models import resnet50
import numpy as np
import paddle 
from paddle.autograd import grad

paddle_model = resnet50(pretrained=True)
paddle_model.eval()

r = np.random.random((1, 3, 32, 32))
random_image = paddle.to_tensor(r).cast("float32")
l = paddle_model(random_image)

# print(grad(l[0], paddle_model.parameters()))  # set only_inputs=True does not help.

for v in paddle_model.parameters():
    print("{}, {}".format(v.name, v.stop_gradient))
target = []
for n in paddle_model.parameters():
    if not n.stop_gradient:
        target.append(n) 
grad(l[0], target)  # set only_inputs=True does not help.
print("Success")
```
则可以正确运行哈：
![image](https://user-images.githubusercontent.com/22361972/168024398-d78b1e64-41d6-475e-bc86-64c9b7f2c16e.png)

- 对于第二个case中，在develop分支已经不会出现问题
```
from paddlenlp.transformers import ErnieForSequenceClassification, ErnieTokenizer
import numpy as np
import paddle
from paddle.autograd import grad

MODEL_NAME = "ernie-1.0"
model = ErnieForSequenceClassification.from_pretrained(MODEL_NAME, num_classes=32)
random_inputs = paddle.to_tensor(np.array([[0]*200]))
l = model(random_inputs, random_inputs)
grad(l, model.parameters())# set only_inputs=True does not help.
print("Success")
```
commit id: e906eb5
![image](https://user-images.githubusercontent.com/22361972/168022798-ba5156fe-45e6-469d-b6d4-6e73c7cb1071.png)

**Comment by holyseven** [comment]
棒！多谢！

抱歉，我才发现resnet50的`parameters()`有bn的mean和variance。

第二个case，在更新到2.3.0之后，也没有类似问题了。
